src/storage/src/client/read_object.rs

```diff
@@ -367,16 +394,92 @@ impl ReadObjectResponse {
     /// # Ok::<(), anyhow::Error>(()) });
     /// ```
     pub async fn next(&mut self) -> Option<Result<bytes::Bytes>> {
-        self.inner.chunk().await.map_err(Error::io).transpose()
+        let res = self.inner.chunk().await.map_err(Error::io);
+        match res {
+            Ok(Some(chunk)) => {
+                self.crc32c = crc32c::crc32c_append(self.crc32c, &chunk);
+                Some(Ok(chunk))
+            }
+            Ok(None) => {
+                let res = check_crc32c_match(self.crc32c, self.check_crc32c());
+                match res {
+                    Err(e) => Some(Err(e)),
+                    Ok(()) => None,
+                }
+            }
+            Err(e) => Some(Err(e)),
+        }
     }
 
     #[cfg(feature = "unstable-stream")]
     #[cfg_attr(docsrs, doc(cfg(feature = "unstable-stream")))]
     /// Convert the response to a [futures::Stream].
     pub fn into_stream(self) -> impl futures::Stream<Item = Result<bytes::Bytes>> {
+        // TODO(#2049): implement checksum checks for streams.
         use futures::TryStreamExt;
         self.inner.bytes_stream().map_err(Error::io)
     }
+
+    fn check_crc32c(&self) -> Option<u32> {
+        // Check the CRC iff all of the following hold:
+        // 1. We requested the full content (request.read_limit = 0, request.read_offset = 0).
+        // 2. We got all the content (status != PartialContent).
+        // 3. The server sent a CRC header.
+        // 4. The http stack did not uncompress the file.
+        // 5. We were not served compressed data that was uncompressed on download.
+        //
+        // For 4, we turn off automatic decompression in reqwest::Client when we create it,
+        // so it will not be turned on.
+        check_crc32c_helper(
+            self.full_content_requested,
+            self.inner.status(),
+            self.inner.headers(),
+        )
+    }
+}
+
+/// Represents an error that can occur when reading response data.
+#[derive(thiserror::Error, Debug, PartialEq)]
+#[non_exhaustive]
+enum ReadError {
```

Comment:
This is private, should we make the error public eventually? I am not sure what is the right answer yet.